week-1-ts-ecosystem/w01e03-object-oriented-structure/src/Cart.ts

```diff
@@ -0,0 +1,53 @@
+import { AuctionProduct, BuyNowProduct, FreeProduct } from "./Products";
+
+type ProductTypes = BuyNowProduct | FreeProduct | AuctionProduct;
+
+export class Cart<Product extends ProductTypes> {
+  #products = new Map<string, Product>();
+
+  getById(id: string) {
+    return this.#products.get(id);
+  }
+
+  addProduct(product: Product) {
+    if (this.#products.has(product.id)) {
+      throw new Error("Product already exists");
+    }
+    this.#products.set(product.id, product);
+  }
+
+  updateProduct(product: Product) {
+    if (!this.#products.has(product.id)) {
+      throw new Error("Product does not exists");
+    }
+    this.#products.set(product.id, product);
+  }
+
+  removeProduct(id: string) {
+    this.#products.delete(id);
+  }
+
+  get products() {
+    return Array.from(this.#products.values());
+  }
+
+  get productCount() {
+    return this.#products.size;
```

Comment:
tutaj chyba jest błąd bo ilość produktów w koszyku to zliczenie wszystkich `.amount` każdego produktu ?